sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java

```diff
@@ -149,49 +191,153 @@ void startLoadJob(
       do {
         try {
-          JobStatus status = client.jobs().get(projectId, jobId).execute().getStatus();
+          Job job = client.jobs().get(projectId, jobId).execute();
+          JobStatus status = job.getStatus();
           if (status != null && status.getState() != null && status.getState().equals("DONE")) {
-            if (status.getErrorResult() != null) {
-              return Status.FAILED;
-            } else if (status.getErrors() != null && !status.getErrors().isEmpty()) {
-              return Status.FAILED;
-            } else {
-              return Status.SUCCEEDED;
-            }
+            return job;
           }
           // The job is not DONE, wait longer and retry.
         } catch (IOException e) {
           // ignore and retry
           LOG.warn("Ignore the error and retry polling job status.", e);
         }
       } while (nextBackOff(sleeper, backoff));
-      LOG.warn("Unable to poll job status: {}, aborting after {} retries.",
-          jobId, MAX_LOAD_JOB_POLL_RETRIES);
-      return Status.UNKNOWN;
+      LOG.warn("Unable to poll job status: {}, aborting after reached max retries.", jobId);
+      return null;
     }
+  }
 
-    /**
-     * Identical to {@link BackOffUtils#next} but without checked IOException.
-     * @throws InterruptedException
-     */
-    private boolean nextBackOff(Sleeper sleeper, BackOff backoff) throws InterruptedException {
+  @VisibleForTesting
+  static class TableServiceImpl implements TableService {
+    private final Bigquery client;
+
+    @VisibleForTesting
+    TableServiceImpl(Bigquery client) {
+      this.client = client;
+    }
+
+    private TableServiceImpl(BigQueryOptions bqOptions) {
+      this.client = Transport.newBigQueryClient(bqOptions).build();
+    }
+
+    @Override
+    public Table getTable(String projectId, String datasetId, String tableId)
+        throws IOException, InterruptedException {
+      BackOff backoff =
+          new AttemptBoundedExponentialBackOff(MAX_RPC_ATTEMPTS, INITIAL_RPC_BACKOFF_MILLIS);
+      return getTable(projectId, datasetId, tableId, Sleeper.DEFAULT, backoff);
+    }
+
+    @VisibleForTesting
+    Table getTable(
+        String projectId,
+        String datasetId,
+        String tableId,
+        Sleeper sleeper,
+        BackOff backoff)
+        throws IOException, InterruptedException {
+      Exception lastException = null;
+      do {
+        try {
+          return client.tables().get(projectId, datasetId, tableId).execute();
+        } catch (IOException e) {
+          LOG.warn("Ignore the error and retry inserting the job.", e);
+          lastException = e;
+        }
+      } while (nextBackOff(sleeper, backoff));
+      throw new IOException(
+          String.format(
+              "Unable to insert job: %s, aborting after %d retries.",
+              tableId, MAX_RPC_ATTEMPTS),
+          lastException);
+    }
+  }
+
+  private static class ReaderImpl implements Reader {
```

Comment:
I am fairly surprised by the number of readers in this PR. What do they all do? Why aren't they all inside of Services classes?